Basic preference experiment action

Osmose opened the pull request:
This adds a new `preference-experiment` action. It has the arguments that we've spec'd out, and forms to add them, but doesn't do anything with them (yet). That's why this targets a feature branch. It also refactors and adds some things to the RecipeForm to support what I wanted to do with the new action.

Because some of the details are still being worked out, I didn't add any tests for this yet. I expect to land most of the features on the feature branch, changing them as I go, and then working on tests as details solidify, but I can change that plan if we think it's a problem. If we're fine with that, then I'll probably merge this if it passes review, regardless of whether the tests pass, and we'll gate the final merge on tests passing.

This covers a fair amount of #587 but doesn't fully resolve it.

r?

andymikulski commented on recipe-server/client/control/components/Fields.js:
```diff
@@ -23,13 +23,14 @@ export function buildControlField({
   label,
   className = '',
   InputComponent,
+  hideErrors = false,
```
Technically the `= false` shouldn't be needed, but I've been a fan of setting defaults this way lately, so I think it's fine anyway.

andymikulski commented on recipe-server/client/control/components/Fields.js:
```diff
@@ -39,10 +40,65 @@ export function buildControlField({
+      type="number"
+      step="1"
+      parse={value => {
+        try {
```
try/catches are slow in JS - could we use underscore's `isNumber` instead? ex:

`...parse={ value => _.isNumber(value) ? parseInt(value, 10) : 0 }...`

mythmon replied:
When would this throw an error anyways? For non-number inputs, parseInt returns `NaN`.

Additionally, won't defining this function inline here cause unneeded re-renders of `ControlField`, since it will have a new value each time?

Osmose replied:
I dunno where I got the idea that `parseInt` throws, but it's been stuck in my head for a long time. Someday I'll get it right.

andymikulski commented on recipe-server/client/control/components/RecipeForm.js:
```diff
@@ -215,6 +225,18 @@ export function initialValuesWrapper(Component) {
+
+    // If we still don't have initial values, roll with the defaults.
+    if (!initialValues) {
+      initialValues = Object.assign({}, DEFAULT_FORM_VALUES);
```
The front end uses the spread operator instead of `Object.assign`, ex:

`initalValues = { ...DEFAULT_FORM_VALUES };`

andymikulski commented on recipe-server/client/control/components/RecipeForm.js:
```diff
@@ -215,6 +225,18 @@ export function initialValuesWrapper(Component) {
+      // ActionField subclasses define their own initial values.
+      if (props.selectedAction) {
+        const ActionFields = RecipeForm.argumentsFields[props.selectedAction];
+        initialValues.arguments = Object.assign({}, ActionFields.initialValues);
```
Spread operator instead of Object.assign

andymikulski commented on recipe-server/client/control/components/action_fields/HeartbeatFields.js:
```diff
@@ -1,99 +1,111 @@
-        type="text"
-      />
+export class HeartbeatFields extends ActionFields {
+  render({ recipeArguments }) {
```
This should default to an empty object, ex `({ recipeArguments = {} })`

Osmose replied:
This does need a default, but `initialValues` on the form is probably a better place for it.

andymikulski commented on recipe-server/client/control/components/Fields.js:
```diff
@@ -39,10 +40,65 @@ export function buildControlField({
+      parse={value => value === 'true'}
+      format={value => {
+        if (value) {
+          return 'true';
```
The booleans-as-strings business seems funky to me, but I'm assuming it's a redux-form thing?

Osmose replied:
Redux-form reads the values from the input element, and input elements return strings as their values. The `parse` and `format` pairs here handle translating between that and a boolean value, which is what I want redux-form to actually store.

andymikulski commented on recipe-server/client/control/sass/control.scss:
```diff
@@ -151,15 +155,32 @@
+  }
+
+  /* Checkbox/radio inputs should appear to the left of the label. */
+  .checkbox-field, .radio-field {
```
Has this been linted?

Osmose replied:
# CORNERS ARE FOR CUTTING

<sub><sup>No, I'll go and fix my linting errors. Sorry.</sup></sub>

Osmose commented:
Feedback'd, this is ready for another pass.

Osmose commented:
Reabsed on master. Pardon the opening and closing, Github was not picking up my changes to the branch for some reason.

andymikulski commented on recipe-server/client/control/components/RecipeForm.js:
```diff
@@ -210,16 +221,32 @@ export const formConfig = {
+        initialValues.arguments = { ...ActionFields.initialValues };
+      } else {
+        // DEFAULT_FORM_VALUES wouldn't copy this, so we add it here instead.
+        initialValues.arguments = {};
```
Instead of `else`ing this, we could probably just do:

```
initialValues = { ...DEFAULT_FORM_VALUES, arguments: {} };
if (selectedAction) { ... }
```

andymikulski commented on recipe-server/client/control/components/action_fields/PreferenceExperimentFields.js:
```diff
@@ -0,0 +1,179 @@
+        <li>
+          <a
+            className="button"
+            onClick={() => fields.push(Object.assign({}, DEFAULT_BRANCH_VALUES))}
```
Use spread operator (`{ ...DEFAULT_BRANCH_VALUES }`)

andymikulski commented on recipe-server/client/control/tests/components/test_RecipeForm.js:
```diff
@@ -33,12 +33,14 @@ describe('<RecipeForm>', () => {
-      );
-      expect(wrapper.find(HeartbeatFields).length).toBe(1);
-    });
+    // TODO(osmose): Figure out how to shallow-render and test connected components
```
I think you're looking for `mount` instead of `shallow`

Osmose replied:
If I use mount, I get an error:

```
Error: Field must be inside a component decorated with reduxForm() in client/control/tests/index.js line 140 > eval (line 59)
```

The shallow render of `RecipeForm` elsewhere works because it's not a connected component (vs the default export, which I don't import during these tests), and the `Field` components never actually get rendered.

I suspect the long-term solution is either use mount and get better at mocking the state tree so that we can fully render connected components during tests, or do some wrapper around the heartbeat fields with a classname we can hook into. The former is definitely the more desirable option, but also conflicts with the upcoming state tree changes.

andymikulski commented on recipe-server/package.json:
```diff
@@ -39,7 +39,7 @@
     "reactable": "0.14.1",
     "redux": "3.6.0",
-    "redux-form": "6.1.0",
+    "redux-form": "6.5.0",
```
<img src="https://cloud.githubusercontent.com/assets/2767162/24172925/ad65ad80-0e4f-11e7-8a41-5e09f66448d9.jpg" height="100" />

Osmose replied:
So it _was_ in a PR!

I'm the worst.

Osmose commented:
Updated based on feedback. I also realized I was doing the function-comparison thing and refactored my components to be class-based components so I could use member functions to avoid the issue.
